tracing/src/main/java/com/palantir/tracing/Tracer.java

```diff
@@ -314,6 +337,37 @@ static boolean isSampled(DetachedSpan detachedSpan) {
         return detachedSpan instanceof SampledDetachedSpan;
     }
 
+    private static final class TraceRestoringCloseableSpanWithMetadata<T> implements CloseableSpan {
+
+        @Nullable
+        private final Trace original;
+
+        private final TagTranslator<? super T> translator;
+        private final T data;
+
+        static <T> CloseableSpan of(@Nullable Trace original, TagTranslator<? super T> translator, T data) {
+            if (original != null || !translator.isEmpty(data)) {
+                return new TraceRestoringCloseableSpanWithMetadata<>(original, translator, data);
+            }
+            return DEFAULT_CLOSEABLE_SPAN;
+        }
+
+        TraceRestoringCloseableSpanWithMetadata(@Nullable Trace original, TagTranslator<? super T> translator, T data) {
+            this.original = original;
+            this.translator = translator;
+            this.data = data;
+        }
+
+        @Override
+        public void close() {
+            Tracer.fastCompleteSpan(translator, data);
+            Trace originalTrace = original;
+            if (originalTrace != null) {
+                Tracer.setTrace(originalTrace);
+            }
```

Comment:
why do we need the `originalTrace` assignment here?

Reply:
oh I see this was just copied from further down in the diff here

Reply:
When we apply a trace somewhere else in a deferred way, it's important that it doesn't destroy any existing tracing that's already happening with a potentially (but not necessarily) different traceId.

Basically the tracing state before this is applied should exactly match the state after it's closed.